core/lib/types/src/storage_writes_deduplicator.rs

```diff
@@ -113,40 +123,65 @@ impl StorageWritesDeduplicator {
                         panic!("tried removing key: {:?} before insertion", key)
                     });
                     *field_to_change -= 1;
+                    *total_size -= value.size;
+                    if is_write_initial {
+                        *total_size -= BYTES_PER_DERIVED_KEY as usize;
+                    } else {
+                        *total_size -= BYTES_PER_ENUMERATION_INDEX as usize;
+                    }
                     updates.push(UpdateItem {
                         key,
                         update_type: UpdateType::Remove(value),
                         is_write_initial,
                     });
                 }
                 (true, Some(new_value)) => {
+                    let mut value_size =
+                        compress_with_best_strategy(log.log_query.read_value, new_value).len();
                     let old_value = self
                         .modified_key_values
-                        .insert(
-                            key,
-                            ModifiedSlot {
-                                value: new_value,
-                                tx_index: log.log_query.tx_number_in_block,
-                            },
-                        )
-                        .unwrap_or_else(|| {
-                            panic!("tried removing key: {:?} before insertion", key)
-                        });
+                        .get(&key)
+                        .unwrap_or_else(|| panic!("key {:?} doesn't exist on update", key));
                     updates.push(UpdateItem {
                         key,
-                        update_type: UpdateType::Update(old_value),
+                        update_type: UpdateType::Update(*old_value),
                         is_write_initial,
-                    })
+                    });
+                    if value_size > old_value.size {
+                        value_size = value_size + old_value.size - 1;
```

Comment:
sorry, why do you reduce `value_size` here? This reduced value will then go to the `ModifiedSlot`

Maybe I misunderstand something, but it seems that a generally easier logic would be:

```
*total_size -= old_value.size;
*total_size += value_size;
```

and that's it

Reply:
the change above covers this but it was because we were only calculating the diff and need to account for the extra metadata byte counted twice. with your comment above this is no longer needed